### PR TITLE
Add new functions to OpalGraphics

### PR DESCRIPTION
**Pull request by theiostream**
This pull requests adds new functions to OpalGraphics or fixes existing ones, almost 90% on CGPath.

### CODE REVIEW

File: Source/OpalGraphics/CGPath.m [hunk 1]
```diff
@@ -159,148 +193,395 @@ CGPoint CGPathGetCurrentPoint(CGPathRef path)
...
+	    break;
+	  case kCGPathElementAddLineToPoint:
+	    if (first)
+	      {
+		return 0;
+	      }
+	    total += winding_line(D(last_p), D(pts[0]), D(point));
+	    last_p = pts[0];
+	    break;
+	  case kCGPathElementAddCurveToPoint:
```
Comment (fredkiefer) [inline Source/OpalGraphics/CGPath.m]: Here we are missing the case of kCGPathElementAddQuadCurveToPoint which does not exist for NSBezierPath.

File: Source/OpalGraphics/CGPath.m [hunk 2]
```diff
@@ -159,148 +193,395 @@ CGPoint CGPathGetCurrentPoint(CGPathRef path)
...
-	// Add the segments to the path:
-	for (index = 0; index < segmentCount; index++, thisAngle += angleStep)
-	{
-	  _OPPathAddArcSegment(path, m,
-	  x, y,
-	  r,
-	  thisAngle,
-	  (thisAngle + angleStep));
-	}
+  NSPoint center = NSMakePoint(x, y);
```
Comment (fredkiefer) [inline Source/OpalGraphics/CGPath.m]: As I already wrote I would prefer the CGPoint version here.

File: Source/OpalGraphics/CGPath.m [hunk 3]
```diff
@@ -311,9 +592,87 @@ void CGPathAddArcToPoint(
...
+  CGFloat l, a1, a2;
+  CGPoint p;
+
+  p = CGPathGetCurrentPoint(path);
+ 
+  dx1 = p.x - x1;
+  dy1 = p.y - y1;
+  
+  l= dx1*dx1 + dy1*dy1;
+  if (l <= 0)
```
Comment (fredkiefer) [inline Source/OpalGraphics/CGPath.m:621]: Should we use a shared bit of logic between this function and CGPathAddArc when to lineTo, moveTo or do nothing? Both current versions look suboptimal.
I have to admit, I did not look at the full version of the function and may be totally off here. What I am think of is that when ever we start an arc we first must decide how to get to the first point of it. And this should be properly implemented.

File: Source/OpalGraphics/CGPath.m [hunk 4]
```diff
@@ -511,3 +870,127 @@ void CGPathAddEllipseInRect(
   p2 = CGPointMake(originx + width / 2 + hdiff, originy + height);
   CGPathAddCurveToPoint(path, m, p1.x, p1.y, p2.x, p2.y, p.x, p.y);
 }
+
+CGRect CGPathGetPathBoundingBox(CGPathRef path)
```
Comment (fredkiefer) [inline Source/OpalGraphics/CGPath.m]: My computer graphics lectures were a long time ago but from what I remember your algorithm seems completely wrong. What you do is compute the bounding box of the curve by leaving out some of the control points. This won't give you the actual bounds of the curve. Please have a look at http://floris.briolas.nl/floris/2009/10/bounding-box-of-cubic-bezier/ or other explanations on the web.
What you basically have to do for the arc bits is to compute enough points on the curve to determine an exact bounding box.

**Review by fredkiefer — CHANGES_REQUESTED** [review]
I think most of this is great and all of it a big improvement over the previous implementation. If you don't want to clean up the things I commented on this is also fine for me, but you should add some comments to the code or in the case of the bounding box, call the old version instead.

File: Source/OpalGraphics/CGPath.m (hunk 1 above)
Comment (theiostream) [reply Source/OpalGraphics/CGPath.m]: I'm not sure on how to implement the winding algorithm for quadratic curves; in an update to this pull request I left it as a `FIXME`. Could you point me in any direction?

File: Source/OpalGraphics/CGPath.m (hunk 2 above)
Comment (theiostream) [reply Source/OpalGraphics/CGPath.m]: Fixed.

File: Source/OpalGraphics/CGPath.m (hunk 4 above)
Comment (theiostream) [reply Source/OpalGraphics/CGPath.m]: Fixed.

File: Source/OpalGraphics/CGPath.m (hunk 3 above)
Comment (theiostream) [reply Source/OpalGraphics/CGPath.m:621]: I agree with your concern; if you think both current versions are suboptimal though, could you suggest a "reference" version so I can reimplement both and keep it uniform?

Comment (fredkiefer) [reply Source/OpalGraphics/CGPath.m:621]: Sorry, I had another look at your current code and the decision in CGPathAddArc now looks correct to me. And the logic in ArcToPoint has to be different. Forget this whole comment, I should have looked at the code of the whole file not just the changes.

**Review by fredkiefer — APPROVED** [review]
With the last changes I think this code is ready for a merge. There are still a few missing bits, but these might be added later.

File: Tests/paths.m [hunk 5]
```diff
@@ -7,15 +7,48 @@
...
+  CGPathRef bounding = CGPathCreateWithRect(CGPathGetPathBoundingBox(path), NULL);
+  CGContextAddPath(ctx, bounding);
+
+  CGMutablePathRef quadCurve = CGPathCreateMutable();
+  CGPathMoveToPoint(quadCurve, NULL, 20, 100);
+  CGPathAddQuadCurveToPoint(quadCurve, NULL, 80, 300, 120, 100);
+  CGPathAddQuadCurveToPoint(quadCurve, NULL, 140, 10, 180, 200);
+  CGContextAddPath(ctx, quadCurve);
+  
+  NSLog(@"wtf %@", NSStringFromRect(CGPathGetBoundingBox(quadCurve)));
```
Comment (ivucica) [inline Tests/paths.m]: Please remove/replace the `wtf` :)

**Review by ivucica — CHANGES_REQUESTED** [review]
I will merge once the `wtf` is gone 👍

**Comment by theiostream** [comment]
Removed the `wtf`. Sorry for that.

**Review by ivucica — APPROVED** [review]

**Pull request merged**